### PR TITLE
Performance: optimize the usage of getBlockIndex

### LINKED ISSUE #13349: Columns and InnerBlocks items defaulting to first column in Chrome
**Describe the bug**
When using Columns or any block with InnerBlocks, the items that you add to second columns get shifted to first columns by default if you use the inserter without modifying the content inside the columns.

**To Reproduce**
1. Add column block
2. Add image placeholder to both columns without using keyboard shortcode, but only the Inserter

**Screenshots**

Here's a video: https://www.useloom.com/share/3296a747738645698c21253ffce1af10

**Desktop (please complete the following information):**
 - OS: Tried on Mac and Windows
 - Browser Chrome
 - Version Many different and latest versions

**Additional context**
- This bug was reported to me by few people testing custom Gutenberg blocks that I made. But it was a weird issue so I tried to setup a new WordPress without any plugins and tried to see if the is because of Gutenberg, and turns out it's an issue with the code.

I believe the focus of InnerBlock stays in the first column item if you just click on the inserter.

### PR DESCRIPTION
**Pull request by youknowriad**
refs #11782 closes #13349

When adding blocks or pressing "Enter", the index of the blocks following the insertion point changes which means all the blocks following this block rerender because they have the `order` as a prop.

In this PR, I move the `getBlockIndex` selector usage to the event handlers instead of computing it in `withSelect` and passing it as a prop to the component to avoid heavy rerenders when pressing "Enter".

### CODE REVIEW

File: packages/editor/src/components/block-list/block.js [hunk 1]
```diff
@@ -769,6 +773,7 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps, { select } ) => {
 } );
 
 export default compose(
+	pure,
```
Comment (youknowriad) [inline packages/editor/src/components/block-list/block.js]: This is useful to avoid rerendering all the blocks when we add/remove a block in the block list (parent component).

I was also tempted to replace it with `React.memo` as it may be more optimized than our HoC. It would be good to compare performance of these two.

File: packages/editor/src/components/block-list/index.js [hunk 2]
```diff
@@ -189,7 +189,6 @@ class BlockList extends Component {
 				{ map( blockClientIds, ( clientId, blockIndex ) => (
 					<BlockListBlock
 						key={ 'block-' + clientId }
-						index={ blockIndex }
```
Comment (youknowriad) [inline packages/editor/src/components/block-list/index.js]: I spent 2 hours trying to find why the blocks were rerendering, it turns out we had a useless prop (that was not even used :P)

File: packages/editor/src/components/block-actions/index.js [hunk 3]
```diff
@@ -45,22 +44,18 @@ export default compose( [
...
-			lastSelectedIndex: getBlockIndex( last( castArray( props.clientIds ) ), rootClientId ),
 			isLocked: !! getTemplateLock( rootClientId ),
 			blocks,
 			canDuplicate,
 			rootClientId,
 			extraProps: props,
 		};
 	} ),
-	withDispatch( ( dispatch, props ) => {
+	withDispatch( ( dispatch, props, { select } ) => {
```
Comment (gziolo) [inline packages/editor/src/components/block-actions/index.js]: @aduth - I don't see it on the list in #12890.

Anyways, I'm happy seeing that `select` is helping us to improve performance. We still didn't discover any downsides which is encouraging to do more refactoring like this.

Comment (aduth) [reply packages/editor/src/components/block-actions/index.js]: >@aduth - I don't see it on the list in #12890.

I think it was missed because I was explicitly looking for `withDispatch( ( dispatch, ownProps )` or `withDispatch( ( dispatch, {`. We apparently have more variation 😬

File: packages/editor/src/components/inserter/index.js [hunk 4]
```diff
@@ -100,27 +100,19 @@ class Inserter extends Component {
...
-			( { rootClientId, index } = insertionPoint );
-		}
+		rootClientId = rootClientId || getBlockInsertionPoint().rootClientId;
 
 		return {
 			title: getEditedPostAttribute( 'title' ),
 			hasItems: hasInserterItems( rootClientId ),
-			rootClientId,
-			index,
+			rootClientId: rootClientId || getBlockInsertionPoint().rootClientId,
```
Comment (aduth) [inline packages/editor/src/components/inserter/index.js]: Should this have been changed from `rootClientId,` ? There's redundancy here in that the same fallback behavior occurs a few lines above.

File: packages/editor/src/components/inserter/menu.js [hunk 5]
```diff
@@ -366,24 +365,54 @@ export default compose(
...
 	} ),
-	withDispatch( ( dispatch, ownProps ) => {
+	withDispatch( ( dispatch, ownProps, { select } ) => {
 		const {
 			__experimentalFetchReusableBlocks: fetchReusableBlocks,
 			showInsertionPoint,
 			hideInsertionPoint,
 		} = dispatch( 'core/editor' );
 
+		// eslint-disable-next-line no-restricted-syntax
```
Comment (aduth) [inline packages/editor/src/components/inserter/menu.js]: We should have a "Disable reason" if only to clarify what's actually being disabled here, given that `no-restricted-syntax` is an umbrella rule containing many miscellaneous custom rules.

File: packages/editor/src/components/block-actions/index.js (hunk 3 above)
Comment (aduth) [reply packages/editor/src/components/block-actions/index.js]: > I think it was missed because I was explicitly looking for `withDispatch( ( dispatch, ownProps )` or `withDispatch( ( dispatch, {`. We apparently have more variation 😬

This is the only occurrence of the second argument being named as "props" for `withDispatch`, and best I could tell (by a regular expression search of `withDispatch\( \( dispatch, [^{o]`) there are no other variations.

**Comment by youknowriad** [comment]
Rebased this, can I get some reviews?

**Comment by youknowriad** [comment]
Updated to fix the tests and this actually also fixes #13349 :)

File: packages/editor/src/components/block-drop-zone/index.js [hunk 6]
```diff
@@ -154,10 +156,11 @@ export default compose(
 		};
 	} ),
 	withSelect( ( select, { rootClientId } ) => {
-		const { getClientIdsOfDescendants, getTemplateLock } = select( 'core/editor' );
+		const { getClientIdsOfDescendants, getTemplateLock, getBlockIndex } = select( 'core/editor' );
 		return {
 			isLocked: !! getTemplateLock( rootClientId ),
 			getClientIdsOfDescendants,
+			getBlockIndex,
```
Comment (gziolo) [inline packages/editor/src/components/block-drop-zone/index.js]: I assume that it isn't something we can handle using `withDispatch` and `select` call.

File: packages/editor/src/components/block-drop-zone/index.js [hunk 7]
```diff
@@ -101,6 +102,7 @@ class BlockDropZone extends Component {
 			return;
 		}
 
+		const dstIndex = dstClientId ? getBlockIndex( dstClientId, dstRootClientId ) : undefined;
```
Comment (gziolo) [inline packages/editor/src/components/block-drop-zone/index.js]: What `dst` stands for? It would read better if we would avoid such abbreviations.

File: packages/editor/src/components/inserter/menu.js [hunk 8]
```diff
@@ -369,24 +368,68 @@ export default compose(
...
+		// it's fine to extract it.
+		// eslint-disable-next-line no-restricted-syntax
+		function getInsertionPoint() {
+			const {
+				getBlockIndex,
+				getBlockRootClientId,
+				getBlockSelectionEnd,
+				getBlockOrder,
+			} = select( 'core/editor' );
+			const { clientId, rootClientId, isAppender } = ownProps;
```
Comment (aduth) [inline packages/editor/src/components/inserter/menu.js]: I don't really understand the purpose of this new `isAppender` prop. Isn't it enough to assume that if the inserter is not provided a `clientId` and there's no current selection, it should take the appending behavior to the `rootClientId`?

Comment (aduth) [reply packages/editor/src/components/inserter/menu.js]: Okay, looking more, I think I understand it a bit better, particularly when the last block is not a paragraph and you have a selection, it's important that the inserter which is shown at the end of content would insert by appending, not after whatever selection happens to exist. The alternative would be that the default appender at the end of content provides the `clientId` of the last block, but I think the expressiveness of this new prop is preferable.

**Review by gziolo — COMMENTED** [review]
This is a quite complex PR. I reviewed it but I was not able properly validate changes in packages/editor/src/components/inserter/menu.js file. I wIll try again after I play with PR in the browser.

Everything else seems to be correct. It's hard to follow as there are some changes in the props passed.

File: packages/editor/src/components/inserter/menu.js [hunk 9]
```diff
@@ -369,24 +368,68 @@ export default compose(
...
+			// If the clientId is defined, we insert at the position of the block.
+			if ( clientId ) {
+				return {
+					index: getBlockIndex( clientId, rootClientId ),
+					rootClientId,
+				};
+			}
+
+			// If there a selected block, we insert after the selected block.
+			const end = getBlockSelectionEnd();
```
Comment (aduth) [inline packages/editor/src/components/inserter/menu.js]: `getBlockSelectionEnd` is performant enough, but there are optimization opportunities here to avoid calling it when `isAppender` is true (allow the [short-circuiting](https://en.wikipedia.org/wiki/Short-circuit_evaluation) of the following condition's `! isAppender`).

```js
let end;
if ( ! isAppender && ( end = getBlockSelectionEnd() ) ) {
	const selectedBlockRootClientId = getBlockRootClientId( end ) || undefined;
	return {
		index: getBlockIndex( end, selectedBlockRootClientId ) + 1,
		rootClientId: selectedBlockRootClientId,
	};
}
```

Or:

```js
if ( ! isAppender ) {
	const end = getBlockSelectionEnd();
	if ( end ) {
		const selectedBlockRootClientId = getBlockRootClientId( end ) || undefined;
		return {
			index: getBlockIndex( end, selectedBlockRootClientId ) + 1,
			rootClientId: selectedBlockRootClientId,
		};
	}
}
```

**Review by aduth — APPROVED** [review]
Looks good. Nice improvements :+1:

This was difficult to review because it wasn't easy to associate individual changes with their goal, as there appears to be about a dozen separate goals of this pull request.

File: packages/editor/src/components/inserter/menu.js (hunk 9 above)
Comment (aduth) [reply packages/editor/src/components/inserter/menu.js]: Oh, I forgot this is only called on specific user interactions, not in a `withSelect`, so not nearly as big a deal.

**Comment by gziolo** [comment]
> Updated to fix the tests and this actually also fixes #13349 :)

I can confirm that it fixes #13349! Nice one :)

File: packages/editor/src/components/inserter/menu.js (hunk 8 above)
Comment (youknowriad) [reply packages/editor/src/components/inserter/menu.js]: Also sometimes, there's no block inside a container block with an appender, and the currently selected block is elsewhere on the page, which means the regular flow would insert the new block under the selected block and not inside the container block.

**Review by gziolo — APPROVED** [review]
In my testing, everything works great in comparison to `master`. It's even better because of this bug fixed.

I spent time trying to figure out why there is no appender after the last block is a Paragraph with content, but it seems like a design decision. @jasmussen is it explained somewhere?

File: packages/editor/src/components/inserter/menu.js (hunk 8 above)
Comment (jasmussen) [reply packages/editor/src/components/inserter/menu.js]: > Technically speaking, I'd love to remove the behavior we have where we insert a block after the selected block, it could be considered confusing and there's the in-between inserters to solve the same use-case.

Followup question to this: if you used the Add Block button in the top toolbar on a very long page, where would the block be inserted? If the answer is "at the end", it doesn't seem like a good solution. Performance is important, of course, but is there any other way we can do it? 

Most word processors insert non-text content where the caret is at. Honestly with the sibling inserter is a new solution with limited prior art. I understand it's a different beast, but we should still not trivially remove this. @alexislloyd what are your thoughts?

Comment (youknowriad) [reply packages/editor/src/components/inserter/menu.js]: I was asking for thoughts and I have them, thanks @jasmussen :) 

My impression was that sometimes, the behavior was not well-understood/confusing for some people but I totally agree that the technically matters shouldn't dictate the decision here, as it's not a big drawback to support this.

**Comment by designsimply** [comment]
> I spent time trying to figure out why there is no appender after the last block is a Paragraph with content, but it seems like a design decision. @jasmussen is it explained somewhere?

@gziolo perhaps this? https://github.com/WordPress/gutenberg/issues/3078#issuecomment-349942072

**Comment by gziolo** [comment]
It's slightly older I think but thanks for checking :)
